Add the reworked API Versioning chapter + annotations in ChangeLog (#273)
admin-shell-io/aas-specs-api: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc

== API Versioning

The versioning scheme for AAS API related services follows semantic versioningfootnote:[http://semver.org]. Very briefly, this defines version numbers as a format following: <MAJOR>.<MINOR>.<PATCH>.

The major version changes in case of breaking or incompatible changes that need to be addressed by clients. Minor versions add (new) functionality in a backwards compatible way and allow clients with lower minor versions to keep their existing functionality. Patch versions only include backwards compatible bug fixes.

AAS API versioning uses the major and minor version as described above. A specific AAS API version uses specific related versions of the metamodel as defined in Clause 1.2. AAS API versions with the same major version must remain compatible, i.e. a client written for an older or a newer minor version must still work. This requires corresponding testing of clients and servers.

> Additionally, "Release candidates" are variants of the implementation of the denoted major version. For example, "3.1.0 RC2" should be interpreted as the second (alternative) release candidate for version 3.1.0.

Reviewer: ```suggestion
Additionally, "Release candidates" are variants of the implementation of the denoted major version. For example, "3.1.0RC2" should be interpreted as the second (alternative) release candidate for version 3.1.0.
```

Reviewer: we did not use a blank before "RC"

> As multiple versions will be supported in the future, an AAS ecosystem consisting of Registry / Discovery services as well as AAS Repository, Submodel (standalone), or AAS (standalone) services should share a consistent version. Therefore, a consistent interface description in the form of OpenAPI documents shall be provided with each major version.

Reviewer: by whom? the openAPI documents are part of the specification itself, aren't they?
I would remove the complete sentence.

> Upcoming compatibility constraints regarding newer versions will be elaborated in further iterations of this document and related technical descriptions (OpenAPI specification).

Reviewer: add in outlook section but not in text part: will typically be forgotten. And I do not know exactly which constraints we do have in mind?

API versioning provides a way to deal with different versions of the same API at the same time. This way, older versions may still be accessible on the same server to provide services to legacy clients without breaking existing functionality. 

> There are different solutions regarding API versioning involving URL-based versioning, query parameter-based versioning, as well as HTTP header-oriented solutions using custom or standard headers.

Reviewer: what is query parameter-based versioning? You have a source for the three classifications?

Author: These are old statements, I have only put them at a different position in the subchapter to keep it readable.

.Example of an AAS endpoint with URL-based versioning

> image::image23.png[Ein Bild, das Dunkel enthält. Automatisch generierte Beschreibung,width=481,height=110]

Reviewer: ```suggestion
image::image23.png[width=481,height=110]
```

As different solutions also provide different advantages and disadvantages, URL-based versioning is recommended for the AAS API. Among other advantages, implementation complexity on clients as well as servers is rather low and different versions can be easily accessed through browsers without the need for specific development tools or extensions.

> Upcoming implementations of AAS related servers may implement the version prefix “api/v<X>/” to provide information of the specific major version regarding AAS Part 2 version, where <X> denotes the implemented major version, e.g. “api/v3/” (see Figure 4). As minor releases of one major version must not contain any breaking changes, the declaration of the minor version can be omitted.

Reviewer: what do you mean? whose impelmentations? the aasx server in Eclipse Digital Twin?

> Nevertheless, AAS servers may decide to use different paths depending on their context, see Figure 5. A client shall not assume that the pattern from Figure 4 is supported by all servers.

Reviewer: ```suggestion
Nevertheless, AAS servers may decide to use different paths depending on their context, see Figure 5. The fragment before the functional endpoint is not necessarily the version information itself, it may also be a tenant ID or any other information the server decides to use.  A client shall not assume that the pattern from Figure 4 is supported by every server.
```

Author: I added the sentence.

.Example of an AAS endpoint with an arbitrary path

> image::image88.png[Ein Bild, das Dunkel enthält. Automatisch generierte Beschreibung,width=481,height=110]

Reviewer: ```suggestion
image::image88.png[width=481,height=110]
```

> Upcoming implementations of AAS related servers may implement the version prefix “*api/v<X.Y>/*” to provide information of the specific major version regarding AAS Part 2 version, where <X> denotes the implemented major version and <Y> denotes the minor version, e.g. “api/v3.0/” (see Figure 4).

Reviewer: why mentioning this if we recommend to use V3 only?

Author: This paragraph is from the old version of the chapter, I have removed it.

> Finally, it is recommended to include an additional "/description” endpoint into each service to further denote information about APIs / servers capabilities. This endpoint provides further information about the API and its supported profiles, and in particular the version of the profiles. The “/description” will be extended with additional information in later versions.

Reviewer: it is not recommended, our profile definitions request it

Reviewer: Future ==> Summary and Outlook, please not in text (it typcially stays forever...)

====

> Note: The profiles provided at the "/description" endpoint (see Clause 10.2.6) contain both the major and minor version declaration.

Reviewer: ```suggestion
Note: The profile identifiers provided at the "/description" endpoint (see Clause 10.2.6) contain both the major and minor version declaration.
```
